library/ceph_crush.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/python
+
+#
+# Copyright (c) 2018 Red Hat, Inc.
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+module: ceph_crush
+
+author: Sebastien Han <seb@redhat.com>
+
+short_description: Create Ceph CRUSH hierarchy
+
+version_added: "2.6"
+
+description:
+    - By using the hostvar variable 'osd_crush_location'
+    ceph_crush creates buckets and places them in the right CRUSH hierarchy
+
+options:
+    cluster:
+        description:
+            - The ceph cluster name.
+        required: false
+        default: ceph
+    location:
+        description:
+            - osd_crush_location dict from the inventory file. It contains
+            the placement of each host in the CRUSH map.
+        required: true
+    containerized:
+        description:
+            - Weither or not this is a containerized cluster. The value is
+            assigned or not depending on how the playbook runs.
+        required: false
+        default: None
+'''
+
+EXAMPLES = '''
+- name: configure crush hierarchy
+  ceph_crush:
+    cluster: "{{ cluster }}"
+    location: "{{ hostvars[item]['osd_crush_location'] }}"
+    containerized: "{{ docker_exec_cmd }}"
+  with_items: "{{ groups[osd_group_name] }}"
+  when:
+    - crush_rule_config
+'''
+
+RETURN = '''#  '''
+
+from ansible.module_utils.basic import AnsibleModule
+import datetime
+
+
+def fatal(message, module):
+    '''
+    Report a fatal error and exit
+    '''
+    if module:
+        module.fail_json(msg=message, rc=1)
+    else:
+        raise(Exception(message))
+
+
+def generate_cmd(cluster, subcommand, bucket, bucket_type, containerized=None):
+    '''
+    Generate command line to execute
+    '''
+    cmd = [
+        'ceph',
+        '--cluster',
+        cluster,
+        'osd',
+        'crush',
+        subcommand,
+        bucket,
+        bucket_type,
+    ]
+    if containerized:
+        for piece in reversed(containerized.split()):
+            cmd.insert(0, piece)
+    return cmd
+
+
+def sort_osd_crush_location(location, module):
+    '''
+    Sort location tuple
+    '''
+    if len(location) < 2:
+        fatal("You must specify at least 2 buckets.", module)
+    [item for item in location if item[0] == "host"] or fatal("You must specify a 'host' bucket.", module)
```

Comment:
A small style nit, I'd probably expand this out to help with readability and make it a bit easier to understand. I think what you're checking here is if the first entry in location looks something like ('host', 'foo'), is that right?

```
if "host" not in location[0]:
    fatal("You must specify a 'host' bucket.", module)
```

Reply:
This is actually looping over the tuple and if no `host` is found then we fail.

Reply:
Let me send a clearer version.

Reply:
A good way to write this without iterating over all the `location` variable would be:

```python
if not any(item for item in location if item[0] == "host"):
    fatal("…")
```

That's a bit more readable and `any()` will cease the iteration as soon as one of the `item` is equal to `"host"`.

Reply:
Thanks @jd, I'll do that.